sql/src/main/java/io/crate/planner/Planner.java

```diff
@@ -109,59 +165,91 @@ public UUID jobId() {
-                            jobSearchContextIdToShard.put(jobSearchContextId, new ShardId(entry.getKey(), shardId));
-                            jobSearchContextIdToNode.put(jobSearchContextId, nodeId);
-                            jobSearchContextId++;
+            for (Map.Entry<String, Map<String, List<Integer>>> location : locations.entrySet()) {
+                for (Map.Entry<String, List<Integer>> nodeRouting : location.getValue().entrySet()) {
+                    Map<Integer, String> shardsOnIndex = shardNodes.get(nodeRouting.getKey());
+                    if (shardsOnIndex == null) {
+                        shardsOnIndex = new HashMap<>(nodeRouting.getValue().size());
+                        shardNodes.put(nodeRouting.getKey(), shardsOnIndex);
+                        for (Integer id : nodeRouting.getValue()) {
+                            shardsOnIndex.put(id, location.getKey());
+                        }
+                    } else {
+                        for (Integer id : nodeRouting.getValue()) {
+                            String allocatedNodeId = shardsOnIndex.get(id);
+                            if (allocatedNodeId != null) {
+                                if (!allocatedNodeId.equals(location.getKey())) {
+                                    success = false;
+                                }
+                            } else {
+                                shardsOnIndex.put(id, location.getKey());
+                            }
                         }
                     }
                 }
             }
+            return success;
         }
 
-        @Nullable
-        public ShardId shardId(int jobSearchContextId) {
-            return jobSearchContextIdToShard.get(jobSearchContextId);
-        }
-
-        public IntObjectOpenHashMap<ShardId> jobSearchContextIdToShard() {
-            return jobSearchContextIdToShard;
-        }
+        public Routing allocateRouting(TableInfo tableInfo, WhereClause where, @Nullable String preference) {
 
-        @Nullable
-        public String nodeId(int jobSearchContextId) {
-            return jobSearchContextIdToNode.get(jobSearchContextId);
+            Collection<TableRouting> existingRoutings = tableRoutings.get(tableInfo.ident());
+            // allocate routingnodes only if we have more than one table routings
+            Routing routing;
+            if (existingRoutings.isEmpty()) {
+                routing = tableInfo.getRouting(where, preference);
+            } else {
+                for (TableRouting existing : existingRoutings) {
+                    assert preference == null || preference.equals(existing.preference);
+                    if (Objects.equals(existing.where, where)) {
+                        return existing.routing;
+                    }
+                }
+                // ensure all routings of this table are allocated
+                for (TableRouting existingRouting : existingRoutings) {
```

Comment:
the stuff here, including the rebuildRouting and the whole NodeRewritingVisitor doesn't have any coverage at all. Can we either write tests for it or replace it with an UnsupportedOperationException until it is actually needed?